earthkit/data/decorators.py

```diff
@@ -35,6 +35,26 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+def detect_out_filename(func):
+    @functools.wraps(func)
+    def wrapped(self, *args, **kwargs):
+        if len(args) == 0:
+            try:
+                args = [os.path.basename(self.path)]
+            except AttributeError:
+                raise TypeError("Please provide a output filename")
+
+        if hasattr(self, "path") and args[0] == self.path:
```

Comment:
This check is not enough. Consider this:
```python
ds = from_source("file", "./my.grib")
ds.save()
```
Now args[0]="my.grib" and self.path="./my_grib". So they are different but point to the same file.

Reply:
good point, I've change to os.path.samefile instead.

Reply:
But given the other point on here, I think this may need to change again when I address that the self.path is the cache hash, not the source filename we want. So I am going to work on that angle before coming back to any other changes here.

Reply:
This is addressed by using os.path.samefile